Introduction of RangeIndex

ARF1 opened the pull request:
closes #939

`RangeIndex(1, 10, 2)` is a memory saving alternative to `Index(np.arange(1, 10,2))`: c.f. #939.

This re-implementation is compatible with the current `Index()` api and is a drop-in replacement for `Int64Index()`. It automatically converts to Int64Index() when required by operations.

At present only for a minimum number of operations the type is conserved (e.g. slicing, inner-, left- and right-joins). Most other operations trigger creation of an equivalent Int64Index (or at least an equivalent numpy array) and fall back to its implementation.

This PR also extends the functionality of the `Index()` constructor to allow creation of `RangeIndex()` with

```
Index(20)
Index(2, 20)
Index(0, 20, 2)
```

in analogy to

```
range(20)
range(2, 20)
range(0, 20, 2)
```

---

TODO:
- [x] Cache a private `Int64Index` object the first time it or its values are required.
- [x] Restore `Index(5)` as error. Restore its test. Allow `Index(0, 5)` and `Index(0, 5, 1)`.
- [x] Make `RangeIndex` immutable. See `start`, `stop`, `step` properties.
- [x] In `test_constructor()`: check class, attributes (possibly including dtype).
- [x] In `test_copy()`: check that copy is not identical (but equal) to the existing.
- [ ] In `test_join_non_unique()`: test .append for ri.append(int64index) and int64index.append(ri).
- [x] In `test_duplicates()`: Assert `is_unique` and `has_duplicates` return correct values.
- [ ] After consensus on desired behaviour, adjust `test_view()`.
- [ ] After consensus on constructor, adjust `test_constructor()` and possible add further tests.
- [x] After consensus on `RangeIndex()` validity, adjust `test_pickle_compat_construction()`.
- [ ] After consensus on slice behaviour, test for correct return type in `test_slice_specialised()`.
- [ ] Accept `xrange` objects as input
- [x] Setting `RangeIndex` as the default index.
- [ ] Clean up `Index` constructor: use ABCs and named arguments
- [ ] In `__getitem__()`: replace `np.isscalar` to exclude singe-element numpy arrays
- [ ] Tests for `pandas.io.packers`

In need of consensus:
- [x] If we disallow `Index(10)` should we also disallow `RangeIndex(10)`? That would seem like an unfortunate api break with respect to built-in `range()`,  `xrange()` and numpy `arange()`. _Decision: allow_
- [x] Should `RangeIndex()` be allowed as a constructor alternative to the kludgy `Index(np.empty(0, dtype='...'))`? _Decision: allow_
- [x] Should `RangeIndex` accept array inputs? If so, what should it do with them? Should it try to infer a corresponding `RangeIndex`? What should it do with inputs it cannot infer? _Decision: delegate all array inputs to `Int64Index`._
- [x] Should the `RangeIndex` constructor accept `dtype` as an argument? _Decision: accept, but only for array inputs._
- [x] If so, should it accept non-int64 dtypes as valid input? _Decision: Irrelevant due to delegation._ 
- [x] If so, does it need to check that the other inputs are within the range of the dtype? _Decision: Irrelevant due to delegation._
- [x] Related to constructor question, should slicing with a list return? Always `Int64Index` or `RangeIndex` when possible and `Int64Index` when not? _Decision: slicing with list will always return `Int64Index`._
- [ ] What should `RangeIndex(10).view('i8')` return? I believe `RangeIndex(10)`. _Decision: return `RangeIndex(10)`._
- [ ] What should `RangeIndex(10).view('i4')` return? I believe `Index(RangeIndex(10), dtype='i4')`. _Decision: return `Index(RangeIndex(10), dtype='i4')`._

Commit:
Introduction of RangeIndex

`RangeIndex(1, 10, 2)` is a memory saving alternative to
`Index(np.arange(1, 10,2))`: c.f. #939.

This re-implementation is compatible with the current `Index()` api and is a
drop-in replacement for `Int64Index()`. It automatically converts to
Int64Index() when required by operations.

At present only for a minimum number of operations the type is
conserved (e.g. slicing, inner-, left- and right-joins). Most other operations
trigger creation of an equivalent Int64Index (or at least an equivalent numpy
array) and fall back to its implementation.

This PR also extends the functionality of the `Index()` constructor to allow
creation of `RangeIndexes()` with
```
Index(20)
Index(2, 20)
Index(0, 20, 2)
```
in analogy to
```
range(20)
range(2, 20)
range(0, 20, 2)
```

Commit:
restore Index() fastpath precedence

jreback commented:
@ARF1 looks like a good start, but need some more testing

- https://github.com/pydata/pandas/blob/master/pandas/core/common.py#L2145 needs to default to a ``RangeIndex``, otherwise the entire point of this change is lost. There *might* be other things like this where ``Int64Index`` are hard-coded. I think need to change them (but look thru them as well)
- need a section in ``tests/test_frame.py``, where you check ``.set_index/reset_index`` and assigning an index. e.g. ``df.index = range(5)`` should tests that its doing the right thing (and work !)
- need a section (make a separate class like I did for ``TestCategoricalIndex`` in ``tests/test_indexing``) to cover various indexing cases. This might be somewhat duplicative, but you can simply cover the most common (e.g.. ``[]/iloc/loc`` with ``scalar/slice/list/index/array`` indexers)
- eventually will need a small doc/warning for 0.17.0 (you can't actually write this atm as it doesn't exist)
- my brief look saw a couple of changed/deleted tests. you *cannot* do that for this kind of non-API change. That is the biggest issue here, that you inadvertantly change some API guarantees in (maybe subtle) ways. That is why we need a lot of tests/validation for this.
- some kind of vbench / metric to see that this indeed a good idea (I think we agree it is, but a couple of simple metrics for showing memory savings / faster ops, whatever). These tend to become part of the whatsnew (like hey pandas changed this and here's why it was done).

put up a list of todo's at the top of the PR, as detailed as possible, and check off as you are done.

We can do an in-depth review once some of these are checked off.

jreback commented on pandas/core/index.py:
```diff
@@ -111,9 +112,31 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None, fastpath=False,
+                    copy = None
+                elif isinstance(copy, int):
+                    range_constructor = True
+                else:
```
this seems very kludgy, what is the issue here?

ARF1 replied:
No real issue. It is not required for `RangeIndex`. I just thought it might be convenient for people to be able to use e.g. `Index(5)` to create a `RangeIndex(5)`.

It is kludgy since it needs to ensure, that the intent of the user really was to create a `RangeIndex` when calling `Index(...)` with a given set of parameters. I don't want to grab inputs that I shouldn't.

If the whole `Index(...)` calling options extension is an issue or just too complicated I am happy to take it out.

jreback replied:
ok, see my other comment. if data is a scalar, then its an error (and btw use ``lib.isscalar``)

shoyer replied:
I would rather we only implement the range like construction when using RangeIndex explicitly. To make a range index from the index constructor, you could write something like `Index(range(...))` (or xrange on Python 2). This could call an alternate constructor, which might just be `_simple_new`. I'll follow up with a more explicit example when I'm not on my phone.

jreback commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+            start = 0
+
+        # check validity of inputs
+        start = cls._ensure_int(start)
```
instead of writing ``_ensure_int``, use ``_ensure_int64`` (from ``core.common``)

jreback commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+    def start(self):
+        return self._start
+
+    @start.setter
```
we don't allow modifying indexes. They are immutable

jreback commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+    def stop(self):
+        return self._stop
+
+    @stop.setter
```
same here

jreback commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+    def step(self):
+        return self._step
+
+    @step.setter
```
same

jreback commented on pandas/tests/test_index.py:
```diff
@@ -297,10 +297,6 @@ def test_constructor(self):
         # self.assertRaises(Exception, arr.view, Index)
 
-    def test_constructor_corner(self):
-        # corner case
```
why did you take this out?

ARF1 replied:
It has to do with the extension of the `Index()` constructor to allow it be used as a short-cut to constructing a `RangeIndex`. This segment tested that e.g. `Index(0)` raised an error since only arrays were valid arguments.

My proposal was that e.g. `Index(5)` could return a `RangeIndex`. The previously invalid constructor use would thus now be valid.

jreback replied:
ok, this changes the API. This is not the pupose of RangeIndex though. It should have *exactly* the same API as ``Int64Index``. So you can allow a fully specified range, e.g. ``RangeIndex(0,10)`` (maybe you allow step to default), but you *cannot* just accept a scalar (I know it seems conivenent), but this is a fundamental property of the ``Index`` type. Pls revert that part of the changes. You can propose that separately (or as a note at the top of the PR), but it must be separate.

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+    def test_constructor(self):
+        index = RangeIndex(5)
+        expected = np.arange(5, dtype=np.int64)
+        self.assert_numpy_array_equal(index, expected)
```
These are not realistic comparisons. You need to assert that they are they correct class and that the attributes are ok (e.g. name) as well.

jreback replied:
furthermore, you allow passing a dtype, so need to check for error on that as well (and to be honest we shouldn't really allow say a ``np.int32`` dtype but we do for a very odd reason in ``Int64Index`` so I guess need to here. But need to tests invalid inputs (as well as valid ones)

ARF1 replied:
Point taken on class & attribute checking.

However `RangeIndex` is currently by definition `int64`. The constructor does/should not allow specification of a `dtype`:
`def __new__(cls, start=None, stop=None, step=None, name=None, fastpath=False)`
Or am I overlooking something?

Or are you saying that I should allow `dtype` specification? I am not sure I am really getting the point here. I could make `dtype` variable. But then I would need to check for all inputs not only that they are `int64` but that they are the right `int` type. Is that worth the effort?

jreback replied:
its that you *need* a dtype. But you need to have the same signature as an ``Int64Index``. because that is how ``Index`` passes things thru. You can raise if its not None and not 'int64', its really for compat only.

The thing is since you can get many types of inputs (e.g. you could get a list/ndarray that is not-convertible to a ``RangeIndex``; should test this and I think raise), but you could also get ``np.arange(...)`` which works (I think I saw some inferernce logic somewhere for this, e.g. given an array determine if its convertible to a RI?)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        self.assertEqual(i_copy.stop, 5)
+        self.assertEqual(i_copy.step, 1)
+        self.assertEqual(i_copy.name, 'Foo')
+
```
need to check that this is not identical (but equal) to the existing

e.g.

``i_copy.equals(i) & not i_copy.identical(i)``

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        self.assertEqual(i_view.name, 'Foo')
+
+        i_view = i.view('i8')
+        tm.assert_index_equal(i, i_view)
```
These don't look right. ``i.view('i8')`` should give you back an ndarray (which ``assert_index_equal`` accepts is equal, but its not the right type, pass ``check_exact=True``).

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        self.assert_numpy_array_equal(index, expected)
+
+        index = RangeIndex(1, 5, 2)
+        expected = np.arange(1, 5, 2, dtype=np.int64)
```
You need a whole bunch of tests that take an ``Index`` and should return a ``RangeIndex`` for certain (e.g. ranges) input, I don't know if these actually exist for ``Int64Index`` (other they would fail).

ARF1 replied:
Not sure I am understanding what you are asking me to do here.

Are you saying that the `RangeIndex(...)` constructor should accept other `Index` types as input, check whether they can be represented as a `RangeIndex` and if so return it? That sounds like a lot of effort for very little gain. What is the use case for something like that?

@shoyer suggested that `RangeIndex` should follow the built-in `range` signature.

jreback replied:
see my point above. I think you need to be pretty compat with what an ``Int64Index`` can accept).

e.g.
- ``range`` operator
- ``np.array`` or list convertible to that could be a ``RangeIndex`` (e.g.. np.arange(5)) falls into this category;
- another ``RangeIndex``
- a ``Int64Index`` convertible as above
- start,stop,step(optional) scalars for describing the range.

You will see for all the index types a fairly flexible constructor.

Start with the ``Int64Index`` constructor

shoyer replied:
The RangeIndex constructor is new. So I really think we restrict acceptable input. The only reason to match Int64Index exactly is if we think users have code like `type(index)(...)` but that looks pretty insane IMO b

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+            joined = self.index.join(self.index, how=kind)
+            self.assertIs(self.index, joined)
+
+    def test_intersection(self):
```
need to test ``.append`` for ``ri.append(int64index)`` and ``int64index.append(ri)``

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        result = idx * idx
+        tm.assert_index_equal(result, didx)
+
+        result = idx / 1
```
fyi currently the numeric operations for mul/div for floats actually do the wrong thing for Int64Index (they return a Int64Index with dtype='float') which is wrong. (outstanding PR to fix this though)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        self.assertRaises(TypeError, lambda : idx * date_range('20130101',periods=5))
+        self.assertRaises(ValueError, lambda : idx * self._holder(3))
+        self.assertRaises(ValueError, lambda : idx * np.array([1,2]))
+
```
so add in numeric operations with the rhs of a float

jreback replied:
nvm you did that below!)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+
+        expected = Float64Index(-arr)
+        a = np.zeros(5,dtype='float64')
+        result = a - fidx
```
add truediv,div,mul tests as well (if you are really adventurous, you can actually put some of these tests in TestsNumeric) (but I suspect you will get some failures for Int64Index at the moment), but better in long run

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+
+    def test_duplicates(self):
+        # RangeIndex has no duplicates by definition
+        pass
```
but still assert it as it has a method for this

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+
+    def test_pickle_compat_construction(self):
+        # RangeIndex() is a valid constructor
+        pass
```
? I think there is a tests for round_trip pickle, very important to tests this

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        # slice value completion
+        index = self.index[:]
+        expected = self.index
+        self.assert_numpy_array_equal(index, expected)
```
again, need to do an index comparison here (with ``check_exact=True`` in this case)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        # positive slice values
+        index = self.index[7:10:2]
+        expected = np.array([14, 18])
+        self.assert_numpy_array_equal(index, expected)
```
this should be returning an ``Int64Index``! so your tests is not specific enough (I don't know whether the code does the right thing)

ARF1 replied:
Which lines exactly do you think should return a `Int64Index`. All slicing? That was my first approach but I decided it was a bad idea since
 1. it would make slicing slower and 
 2. I thought `RangeIndex` should really be conserved for slices: If a user set `RangeIndex` for a df, surely she/he would want it on the slices as well!?

Why do you think slicing should materialise the `RangeIndex` to `Int64Index`?

jreback replied:
no I wouldn't return a ``Int64Index`` with slicing, unless you cann't return a ``RangeIndex``. I think this should be cheap to figure out what you can actually return. 

slices -> ``Rangeindex`` (always?). 
scalar -> scalar
list-like -> if contiguous then -> RI?, otherwise Int64

e.g.

``RangeIndex(0,10)[0,1]`` -> ``RangeIndex(0,1)``
``RangeIndex(0,10)[0,2]`` -> ``Int64Index([0,2])`` (unless you have a good way of figuring this out)
``RangeIndex(0,10)[0,2,3]`` -> ``Int64Index([0,2,3])``

jreback replied:
the particular example I noted *should* return an ``Int64Index``. More the point, ALL of those should return either a ``RangeIndex`` or ``Int64Index``, NOT an ndarray.

shoyer replied:
I think only slicing with a slice object should return another range index.

jreback replied:
I disagree. The most important part of this is to try hard to preserve the ``RangeIndex`` if at all possible. If that's not the goal, then this is just a novelty because you can't expect it to be preserved when ops happen. @shoyer don't you remember all of the discussion w.r.t. ``CategoricalIndex`` about how it should be preserved ad-nauseum? how is this any different? how are Indexes in general any different?

shoyer replied:
A range index doesn't have a special dtype like categorical. We should certainly preserve RangeIndex when possible, but it should be entirely predictable for each operation. It's certainly not predictable if indexing with some lists/arrays returns a RangeIndex and with others returns an Int64Index.

ARF1 replied:
I also think we should preserve `RangeIndex` when possible. Thus, slicing with a slice object should always return a `RangeIndex` in my opinion.

When slicing with an array however, I believe we should always return `Int64Index`. Not however because I fear the unpredictability of sometimes returning `RangeIndex` and sometimes returning `Int64Index`. Since we are aiming for a "drop-in replacement" for `Int64Index` that really should not matter in my opinion.

What concerns me much for is the typical use case of slicing with an array.
@jreback wants `RangeIndex(0, 10)[np.arange(10)]` to return `RangeIndex(0, 10)`. The problem is, that to make this happen, we need to analyse the array provided and check whether this is actually possible. That is a fairly costly operation for something as ubiquitous as indexing/slicing.

In my opinion no sane user will write something like `df[np.arange(0, 10)]` when something like `df[0:10]` will do the trick.

Users will pass an array if they are pulling out elements not in-order but out-of-order. E.g. after an argsort-like operation. With @jreback's suggestion, the operation for these users will be much slower than expected without much gain given the unlikely nature of indexing with a `range`-compatible array.

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2447,6 +2443,567 @@ def test_slice_keep_name(self):
+        self.assert_numpy_array_equal(index, expected)
+
+        # stop overshoot
+        index = self.index[2:100:4]
```
same for the rest of these

shoyer commented:
Here's thought on the constructor:
```python
class Index(IndexOpsMixin, PandasObject):
    # ...

    def __new__(cls, data=None, dtype=None, copy=False, name=None, fastpath=False,
                 tupleize_cols=True, **kwargs):
        # ...

        if all(hasattr(data, a) for a in ['start', 'stop', 'step']):
            return RangeIndex._simple_new(data, name)

# ...

class RangeIndex(Int64Index):
   def __new__(cls, *args, **kwargs):
       # range should be xrange on Python 2
       range_obj = range(*args)
       name = kwargs.pop('name')
       if kwargs:
           raise TypeError('too many keyword arguments')
       return _simple_new(range_obj, name)

    @classmethod
    def _simple_new(cls, range_obj, name=None):
        result = object.__new__(cls)
        result._start = range_obj.start
        result._stop = range_obj.stop
        result._step = range_obj.step
        result.name = name
        return result
```

jreback commented:
I think there are a ton of edge cases that can be dealt with by having a proper constructor. E.g.

``Index(np.arange(5))`` -> *should* return ``RangeIndex``
``Index(range(5))`` -> ``RangeIndex``
``Index([0,1,2,3,4,5])`` -> what does this return?

I don't think there will be much explict usage of ``RangeIndex``, but it should be a drop-in replacement for ``Int64Index``, thus should be able to accept the same exact input

ARF1 commented:
@jreback Thank you very much for taking the time for extensively commenting the code. I really appreciate it.

On some points I will need clarification (see comments in the line notes). Two more fundamental questions I think we should discuss follow below:

### Making `RangeIndex` the default index ###
I held off on making `RangeIndex` the default index class since I thought a "soft phase-in" might be a better approach than an "all or nothing" approach: Once it is in the code base, people could start to use it and remaining bugs could be ironed out before making it the default.

Also, until really all conceivable operations are conserving the `RangeIndex` type, there will probably be performance implications in some cases since materialization to `Int64Index` will be triggered by many operations (probably) which will be slower than the use of an already existing index as is the case currently.

Materialization of the index will happen repeatedly if the user calls non-`RangeIndex`-conserving functions repeatedly on an object indexed with a `RangeIndex`. (See also my line comments regarding the slicing operation.)

In its present form I do not see `RangeIndex` as a performance PR. In the long term I have a hunch that it could be but I think we are still far from it. Making it so, is a far larger effort I believe. Also I am not sure I am really up to that task. - I am kind of surprised I managed to get it working at all as a pandas novice. ;-)

Are you 100% certain you want to make it the default index straight away? I really was trying to keep the PR manageable to start with and optimize its performance with smaller, more manageable subsequent PRs.

### changed/deleted tests
There is only the one `test_constructor_corner(self)` test that I deleted.

This became necessary since I suggested allowing more inputs to the `Index(...)` constructor. If I change that constructor I have to change its tests accordingly. (See also line comments.)

I could restore the `Index(...)` constructor to its current, more limited functionality in which case I can then also restore the test. Should I to that?

shoyer commented:
RE the cost of repeatedly allocating the values -- I would cache the array of values on the RangeIndex once accessed.

jreback commented:
@ARF1 np, always nice to have someone work on the internals!

I think the original intention WAS to have ``RangeIndex`` be the default index created. However as you have noted then the semantics of the operations need to be quite clearly defined, e.g. when and when not to materialize to a baser type (e.g. we experience similar discussion on when to morph ``CategoricalIndex`` to a ``Index``).

@shoyer does have a good point that simple, predictible rules are useful. 

If its release in a more limited scope (e.g. say non-default and slices only preserving, with a similar to existing constructor). Then how/when would this actually be created / where would it be useful enough that one would explicity create it?

See that's the rub. Pandas does lots of inference for users (arguable points whether too much at times of course!).

So if somethings natural (and I think we agree this is), then it should be put in the natural places for utility.

shoyer commented:
@jreback My notion of a drop in replacement is different. I don't think we should allow for `RangeIndex(np.arange(...))` and we *certainly* should not try to infer a range index from `Index(np.arange(...))`. That is simply too magical, and will have unexpected performance implications -- converting an integer ndarray to an Index should take constant, not linear time.

I would like to see this enabled as the default index (when none is explicitly provided) for two reasons:

1. Currently, the default index *is* very wasteful. It often gets created and then immediately thrown away. This happens all the time in user code.
2. Testing RangeIndex as the default will be necessary to be confident that it works properly, and at that point, we've already done most of the hard work here.

ARF1 commented:
@shoyer Regarding your constructor suggestion above: the `xrange` object in python 2 is much more limited than the `range` object in python 3. In particular it does not have `start`, `stop` or `step` properties. We could infer these by testing the return from the object - except for `stop` for which we could at best determine a new value that will lead to the same results.

To be honest, I do not really see the point however. What sane user would write `Index(xrange(0,10))` when `Index(0, 10)` does the same thing?

In short: of course we can do it, but is it worth the effort?

@shoyer Regarding your suggestion of caching the values of `RangeIndex` when they are accessed: I probably misunderstood you here but:

 - If we cache all the values in a `RangeIndex` why not just convert to a `Int64Index` in the first place? With your suggestion we are carrying around two objects describing the same thing that we need to keep synchronized. - Sounds like a nightmare to me.
 - If we cache only parts of the values in the `RangeIndex`: how would that work in practice?

ARF1 commented:
@jreback @shoyer I added a list of topics on which we still need a consensus (or alternatively an authoritative decision) to the OP.

ARF1 commented:
@shoyer Re caching values: Never mind, I was a bit slow. I think I now understand what you meant: For `RangeIndex` to have a sort of lazy instantiation of a private `Int64Index` object the first time the values or a `Int64Index` is required. I am going to implement it and we'll see whether we mean the same thing.

shoyer commented:
For caching values, you can use the `cache_readonly` decorator. So this looks something like:
```
@cache_readonly
def _data(self):
    # ...
```

shoyer commented:
As for handling `range`/`xrange` as an argument:

I think this is nice to support, simply because it's slightly fewer characters to type than `pd.RangeIndex`. It's also nice because it gives us an unambiguous way to use the generic `Index` constructor to make a `RangeIndex`. (I think both @jreback and I have noted that supporting `Index(start, stop, step)` seems like a bad idea.)

It's definitely unfortunate Python 2 doesn't define start, step and stop for `xrange`. But that information is all still accessible via `repr`. So you could write something like:
```
if PY2 and isinstance(data, xrange):
    g, = re.match('xrange\((.+)\)', repr(data)).groups()
    return RangeIndex(*map(int, m.split(', ')))
```

Not so elegant, but it gets the job done.

shoyer commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+        if np.isscalar(key):
+            n = int(key)
+            if n != key:
+                return super_getitem(key)
```
I think this should always `raise KeyError`? Or do we actually allow for non-integer `__getitem__` on a `Int64Index`? If I recall correctly, numpy has deprecated that sort of indexing...

ARF1 replied:
IMHO, this is correct:

```
In [6]: pd.Int64Index([1,2,3,4])[1.0]
Out[6]: 2
```

Another issue is, that `np.isscalar` is too permissive. (It allows single-element numpy arrays.) Probably better to use `isinstance(key, numbers.Integer)` or something similar to replicate numpy `__getitem__` behaviour.

jreback replied:
this is deprecated for all other index types except Float64Index ; we do not allow float indexers for a lot of reasons

ARF1 replied:
Ok, that changes things. Can we then for consistency reject all non-int-type inputs for `RangeIndex`? Especially for the constructor.

shoyer replied:
Yes, please do.

shoyer commented on pandas/core/index.py:
```diff
@@ -3299,6 +3322,415 @@ def _wrap_joined_index(self, joined, other):
+            return self.start + n * self.step
+
+        if isinstance(key, slice):
+            # complete missing slice information
```
Like I said before, I would love to be able to delegate this logic to the (Python 3) `range` object. There are a surprising number of tricky edge cases, especially with things like negative step size:
```
In [1]: x = range(0, 10, 2)

In [2]: x
Out[2]: range(0, 10, 2)

In [3]: x[:2]
Out[3]: range(0, 4, 2)

In [4]: x[::2]
Out[4]: range(0, 10, 4)

In [5]: x[::-1]
Out[5]: range(8, -2, -2)

In [6]: x[3:6]
Out[6]: range(6, 10, 2)
```

shoyer replied:
OK, maybe it's not so tricky after all:
https://github.com/PythonCharmers/python-future/blob/521477b69e4148f720b2f0b6b5287e5947bff203/src/future/types/newrange.py#L112

But at the very least you should copy the test-suite:
https://github.com/PythonCharmers/python-future/blob/521477b69e4148f720b2f0b6b5287e5947bff203/tests/test_future/test_range.py

ARF1 replied:
Thanks for the links. I decided to delegate the entire logic to the `slice` object. (Like they did in `future`.) Its quite convenient since it is available on python2 and python3. Also, it saves me from fixing my broken implementation.

Commit:
Various fixes suggested by @jreback and @shoyer

Cache a private Int64Index object the first time it or its values are required.
Restore Index(5) as error. Restore its test. Allow Index(0, 5) and Index(0, 5, 1).
Make RangeIndex immutable. See start, stop, step properties.
In test_constructor(): check class, attributes (possibly including dtype).
In test_copy(): check that copy is not identical (but equal) to the existing.
In test_duplicates(): Assert is_unique and has_duplicates return correct values.

ARF1 commented:
@jreback @shoyer I spun-off work on the "default index issue" to a new PR (#9999) to avoid messing up this PR with its more limited scope.

Commit:
fix slicing

Commit:
Set RangeIndex as default index
* enh: set RangeIndex as default index
* fix: pandas.io.packers: encode() and decode() for RangeIndex
* enh: array argument pass-through
* fix: reindex
* fix: use _default_index() in pandas.core.frame.extract_index()
* fix: pandas.core.index.Index._is()
* fix: add RangeIndex to ABCIndexClass
* fix: use _default_index() in _get_names_from_index()
* fix: pytables tests
* fix: MultiIndex.get_level_values()
* fix: RangeIndex._shallow_copy()
* fix: null-size RangeIndex equals() comparison
* enh: make RangeIndex.is_unique immutable

Commit:
fix view

Commit:
enh: various performance optimizations

 * optimize argsort()
 * optimize tolist()
 * comment clean-up

ARF1 commented:
I am considering making `RangeIndex` a subclass of `slice` to be able to use a `RangeIndex` for slicing objects without instantiating a numpy array. (Since the `__getitem__()` functions seem to consistently check for instances of `slice` before checking for instances of `ndarray`.)

What do you think?

jreback commented:
no, don't subclass except for ``Index``. This just makes things harder / more confusing.

shoyer commented:
Yeah, that seems likely to be quite problematic. For example, a negative start has a totally different meaning between range (start at that value) and slice (count from the end).

On Thu, May 7, 2015 at 9:39 AM, jreback <notifications@github.com> wrote:

> no, don't subclass except for ``Index``. This just makes things harder / more confusing.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/9977#issuecomment-99932759

auvipy commented:
@ARF1 could u plz fix the merge conflict?

jreback commented:
replace by #11892 
cc @ARF1
